### ISSUE TITLE
v5.2.2: Browser is not asking for authentication with HTTP (proxy chaining with squid) as first request 

### ISSUE THREAD

**Issue by Schroeffu** [e2guardian/e2guardian#464]
I have E2Guardian active with Squid NTLM authentification and MITM (e2guardian). Furthermore I have 4 filter Groups: filter1-3 cannot download .exe files && executable MIME types, filter4 is allowed to do. The following test-case is done with a user in xxxx=filter4

The .exe Extension/MIME Type Filter is working correct on any HTTPS site, filter1-3 is blocked, filter4 can download. But on unencrypted HTTP sites the filter4 is not working. It does block the download with my filter=4 user. Again: In a 2nd tab, test-download an .exe over https, the filter4 is correctly allowing me download the .exe file.

`Test-Download URL HTTP: http://prdownloads.sourceforge.net/filezilla/FileZilla_3.2.7.1_win32-setup.exe`

`Test-Download URL HTTPS: https://download.filezilla-project.org/client/FileZilla_3.37.4_win64-setup_bundled.exe`

Log HTTP download:

```
1540221802.291          -       172.16.5.15     172.16.5.15     http://prdownloads.sourceforge.net/filezilla/FileZilla_3.2.7.1_win32-setup.exe  GET     403     0       -       -       -       0       -       900     *VERWEIGERT* Verbotene Datei-Erweiterung: .exe       0       -       InternetAccessLevel0    1
1540221802.334          -       172.16.5.15     172.16.5.15     http://prdownloads.sourceforge.net/favicon.ico  GET     407     4005    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
```

But in a 2nd tab, download an .exe with HTTPS is working. Don't ask me why there is no username in logfile, another bug? See also https://github.com/e2guardian/e2guardian/issues/441

Log HTTPS download: 
```
1540221925.434          -       172.16.5.15     172.16.5.15     https://download.filezilla-project.org:443      CONNECT 407     3632    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
1540221925.434          -       172.16.5.15     172.16.5.15     https://download.filezilla-project.org:443      CONNECT 407     3632    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
```

For sites on "exceptionsitelist" the user is logged properly. For example I have google.com on this exceptionsitelist whitelisted. Example Log:

```
1540222050.755          schroeffu  172.16.5.15     172.16.5.15     https://sb-ssl.google.com:443   CONNECT 200     4885    -       -       -       120329  -       602     *VERTRAUENSWÜRDIG* Übereinstimmung mit Seite in Ausnahmelistegoogle.com      0       -       InternetAccessLevel3    4
```

If I have to debug the main issue more, can somebody explain me how to enable Debug in the Pre-Compiled Ubuntu 18.04 .deb Version?

Thanks!
Schroeffu

**Comment by fredbcode** [comment]
407 logs are generated by NTLM requests, you must see exactly the same requests in Squid.
"https://wiki.squid-cache.org/ConfigExamples/Authenticate/Ntlm" at this steep there is no username at all, no problem

Yes here there is no username, but something is missing we have here only the first part of the challenge (407), where is the real request we must see a code 200/304 right after.
```

1540221925.434          -       172.16.5.15     172.16.5.15     https://download.filezilla-project.org:443      CONNECT 407     3632    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
1540221925.434          -       172.16.5.15     172.16.5.15     https://download.filezilla-project.org:443      CONNECT 407     3632    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
```

I'm seeing no url at all (just domain) are you sure SSLMITM is enabled and working ? Are you seeing your certificate in your browser and not the usual from website ?

Here in http request we see the url, but there is no username (and no 407) the user is "logued" as filter1 
 (default) and the request is denied, something is wrong with NTLM or we enter in another rule like address IP is in filter1 ?

`1540221802.291          -       172.16.5.15     172.16.5.15     http://prdownloads.sourceforge.net/filezilla/FileZilla_3.2.7.1_win32-setup.exe  GET     403     0       -       -       -       0       -       900     *VERWEIGERT* Verbotene Datei-Erweiterung: .exe       0       -       InternetAccessLevel0    1`
 
NTLM is a very problematic identification method

**Comment by fredbcode** [comment]
http://prdownloads.sourceforge.net/filezilla/FileZilla_3.2.7.1_win32-setup.exe is a "wrong" url with redirection

**Comment by Schroeffu** [comment]
Thanks for very much for your answer. I had loglevel 2 (all text based) instead of 3 (all requests). 

Now the full logs, trying open a HTTP Site with ending .exe: http://prdownloads.sourceforge.net/filezilla/FileZilla_3.2.7.1_win32-setup.exe

```
1540279557.365          -          172.16.5.15     172.16.5.15     https://www.google.com:443      CONNECT 407     3568    -       -       -       0       -       602     *VERTRAUENSWÜRDIG* Übereinstimmung mit Seite in Ausnahmelistegoogle.com      0       -       InternetAccessLevel0    1
1540279557.551          -          172.16.5.15     172.16.5.15     http://prdownloads.sourceforge.net/filezilla/FileZilla_3.2.7.1_win32-setup.exe  GET     403     0       -       -       -       0       -       900     *VERWEIGERT* Verbotene Datei-Erweiterung: .exe       0       -       InternetAccessLevel0    1
1540279557.827          -          172.16.5.15     172.16.5.15     http://prdownloads.sourceforge.net/favicon.ico  GET     407     4103    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
1540279558.552          schroeffu  172.16.5.15     172.16.5.15     http://prdownloads.sourceforge.net/favicon.ico  GET     307     280     -       -       -       527     -       0       *ÜBERPRÜFT* -   0       -       InternetAccessLevel34
1540279558.747          -          172.16.5.15     172.16.5.15     https://sourceforge.net:443     CONNECT 407     3572    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0    1
1540279559.883          schroeffu  172.16.5.15     172.16.5.15     https://sourceforge.net/downloads       GET     301     274     -       -       -       174     -       0       *ÜBERPRÜFT* -   0       -       InternetAccessLevel3    4
1540279560.242          schroeffu  172.16.5.15     172.16.5.15     https://sourceforge.net/publish GET     301     276     -       -       -       198     -       0       *ÜBERPRÜFT* -   0       -       InternetAccessLevel3    4
1540279560.486          schroeffu  172.16.5.15     172.16.5.15     https://sourceforge.net/publish/        GET     301     272     -       -       -       181     -       0       *ÜBERPRÜFT* -   0       -       InternetAccessLevel3    4
1540279560.743          schroeffu  172.16.5.15     172.16.5.15     https://sourceforge.net/create  GET     301     274     -       -       -       179     -       0       *ÜBERPRÜFT* -   0       -       InternetAccessLevel3    4
1540279561.351          schroeffu  172.16.5.15     172.16.5.15     https://sourceforge.net/create/ GET     200     136865  -       -       -       532     -       0       *ÜBERPRÜFT* -   0       -       InternetAccessLevel3    4
```

Open a HTTPS Site with ending .exe works:

```
1540280055.218          -          172.16.5.15     172.16.5.15     https://download.filezilla-project.org:443      CONNECT 407     3632    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
1540280055.482          -          172.16.5.15     172.16.5.15     https://download.filezilla-project.org:443      CONNECT 407     3632    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
1540280082.489          schroeffu  172.16.5.15     172.16.5.15     https://download.filezilla-project.org/client/FileZilla_3.37.4_win64-setup_bundled.exe  GET     200     8784528 -       -       -       26644   -       0       *ÜBERPRÜFT* -0       -       InternetAccessLevel3    4
```

No there is no ipfilter for filter1.
MITM is definitively working, i have test-virus on http & https site which is detected by ICAP Scanning properly on both, http and https:

```
http site test virus:

1540280755.722          -          172.16.5.15     172.16.5.15     http://test.schroeffu.ch/eicar.com.txt  GET     407     3861    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel01
1540280755.999          schroeffu  172.16.5.15     172.16.5.15     http://test.schroeffu.ch/eicar.com.txt  GET     403     68      -       -       -       214     -       0       *INFIZIERT* * *VERWEIGERT* Ein Virus oder unerlaubter Inhalt wurde gefunden. Unknown 0       Content scanning        InternetAccessLevel3    4

https site test virus:

1540280702.126          -          172.16.5.15     172.16.5.15     https://schroeffu.ch:443        CONNECT 407     3560    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0    1
1540280702.167          -          172.16.5.15     172.16.5.15     https://schroeffu.ch:443        CONNECT 407     3560    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0    1
1540280702.513          schroeffu  172.16.5.15     172.16.5.15     https://schroeffu.ch/eicar.com.txt      GET     403     68      -       -       -       81      -       0       *INFIZIERT* * *VERWEIGERT* Ein Virus oder unerlaubter Inhalt wurde gefunden. Unknown 0       Content scanning        InternetAccessLevel3    4
```

Does it make sense to change SQUID + E2Guardian to Basic-Auth only and test with that settings? At the I need NTLM or Kerberos (not supported by e2guardian yet as i know), for SSO on Windows via ActiveDirectory.

**Comment by Schroeffu** [comment]
> 
> 
> http://prdownloads.sourceforge.net/filezilla/FileZilla_3.2.7.1_win32-setup.exe is a "wrong" url with redirection

Yes but we can try as demonstration any site with .exe ending, here is another real existing example which is blocked by HTTP but should not:

```
1540281007.887          -         172.16.5.15     172.16.5.15     http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe     GET     403     0       -       -       -       0       -       900     *VERWEIGERT* Verbotene Datei-Erweiterung: .exe       0       -       InternetAccessLevel0    1
1540281008.041          -         172.16.5.15     172.16.5.15     http://downloads.signotec.com/favicon.ico       GET     407     4070    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
1540281008.299          schroeffu 172.16.5.15     172.16.5.15     http://downloads.signotec.com/favicon.ico       GET     404     209     -       -       -       198     -       0       *ÜBERPRÜFT* -   0       -       InternetAccessLevel34
```

**Comment by fredbcode** [comment]
Yes, because the user is no authenticate there is something wrong here, 
do you have a request right before, nothing ?

Maybe related with storyboard rules order

**Comment by fredbcode** [comment]
||
> |Does it make sense to change SQUID + E2Guardian to Basic-Auth only and 
> test with that settings? At the I need NTLM or Kerberos (not supported 
> by e2guardian yet as i know), for SSO on Windows via ActiveDirectory. |
>

Yes really, please

Don't worry about kerberos I have a magic trick ...

**Comment by Schroeffu** [comment]
Ok i changed E2Guardian + Squid to BasicAuth only. There is the same problem, it seems not to be NTLM based issue. 

After changing to BasicAuth + Restart E2Guardian/Squid, the very first request to this HTTP URL http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe is not asking for any username/password auth, e2guardian just blocks the request according to the entry in lists/bannedextensionlist.

Here the workflow:

-  Change SQUID + E2Guardian to BaisAuth only
-  Open directly a HTTP .exe ending page fails (browser dont ask for username/password) http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe

```
1540281300.166          -       172.16.5.15     172.16.5.15     http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe     GET     403     0       -       -       -       0       -       900     *VERWEIGERT* Verbotene Datei-Erweiterung: .exe       0       -       InternetAccessLevel0    1
1540281300.351          -       172.16.5.15     172.16.5.15     http://downloads.signotec.com/favicon.ico       GET     407     4248    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
```

- Open https://github.com (so, HTTPS) in a 2nd tab is asking username/password in Browser and i can log me in:

```
1540281326.887          -         172.16.5.15     172.16.5.15     https://github.com:443  CONNECT 407     3730    -       -       -       1       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0    1
1540281326.887          -         172.16.5.15     172.16.5.15     https://github.com:443  CONNECT 407     3730    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0    1
1540281345.432          schroeffu 172.16.5.15     172.16.5.15     https://github.com      GET     200     47273   -       -       -       521     -       0       *ÜBERPRÜFT* -   0       -       InternetAccessLevel3    4
1540281345.853          schroeffu 172.16.5.15     172.16.5.15     https://avatars3.githubusercontent.com/u/12001003?s=40&v=4      GET     304     0       -       -       -       155     -       0       -       0       -       InternetAccessLevel3 4
(...)
```

- Open again in 3rd tab the HTTP .exe ending page works now with logged in user http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe:

```
1540281479.152           schroeffu 172.16.5.15     172.16.5.15     http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe     GET     200     28204991        -       -       -   97436    -       0       -       0       -       InternetAccessLevel3    4
1540281602.179           schroeffu 172.16.5.15     172.16.5.15     http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe     GET     206     27189728        -       -       -   93935    -       0       -       0       -       InternetAccessLevel3    4
1540281732.457           schroeffu 172.16.5.15     172.16.5.15     http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe     GET     206     27876320        -       -       -   101152   -       0       -       0       -       InternetAccessLevel3    4
1540281845.285           schroeffu 172.16.5.15     172.16.5.15     http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe     GET     206     28213568        -       -       -   83730    -       0       -       0       -       InternetAccessLevel3    4
1540281887.529           schroeffu 172.16.5.15     172.16.5.15     http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe     GET     206     4791537 -       -       -       13149-       0       *ÜBERPRÜFT* -   0       -       InternetAccessLevel3    4
```

So for me it looks like E2Guardian is not asking for authentification for HTTP Sites. 

My Setting in E2Guardian.conf for this BasicAuth Test:


```
authplugin = '/etc/e2guardian/authplugins/proxy-basic.conf'
#authplugin = '/etc/e2guardian/authplugins/proxy-digest.conf'
#authplugin = '/etc/e2guardian/authplugins/proxy-ntlm.conf'
#authplugin = '/etc/e2guardian/authplugins/ident.conf'
#authplugin = '/etc/e2guardian/authplugins/ip.conf'
#authplugin = '/etc/e2guardian/authplugins/proxy-header.conf'
#authplugin = '/etc/e2guardian/authplugins/port.conf'
```

Squid.conf Basic Auth related:

```
######################################################### NTLM #########################################################
# NTLM authentication, caching not neccessary, because the connection will be authenticated against the token
#auth_param ntlm program /usr/bin/ntlm_auth --helper-protocol=squid-2.5-ntlmssp --use-cached-creds --offline-logon
#auth_param ntlm realm K+A Proxy-Anmeldung NTLM: Bitte Domaene\Benutzername und Windows-Kennwort eingeben.
#auth_param ntlm children 50
#auth_param ntlm keep_alive on

######################################################### BASIC #########################################################
# Basic authentication, "proxy_auth" acl neccessary, group nesting possible
auth_param basic program /usr/bin/ntlm_auth --helper-protocol=squid-2.5-basic
auth_param basic children 50
auth_param basic credentialsttl 5 minutes
auth_param basic realm  Proxy BasicAuth: Bitte Windows-Benutzername und Windows-Kennwort eingeben.
auth_param basic casesensitive off

######################################################### Allow based on group membership #########################################################
# Authentication required, otherwise Pop-Up
acl Authenticated_Users proxy_auth REQUIRED

(...)
```

Hmmm i am confused, does it really affect only me?

**Comment by Schroeffu** [comment]
> 
> Yes, because the user is no authenticate there is something wrong here, do you have a request right before, nothing ? Maybe related with storyboard rules order

Sorry no, before the denied log line there is nothing else in the log related to this request.
Storyboard sounds possible, I didn't change any rule in the existing .story files yet. Do you have an idea which rules there are related to authentification/https/http?

**Comment by Schroeffu** [comment]
By the way: with stopped E2Guardian and only accessing proxy SQUID in BasicAuth Mode, the above mentioned http .exe URL is opening correctly, SQUID is asking immediately for authentification and download of .exe runs after that login. So it must be e2guardian related.

**Comment by fredbcode** [comment]
`So for me it looks like E2Guardian is not asking for authentification for HTTP Sites.`

E2guardian asks nothing, juste read the request and catch the username form squid
Here the problem is that e2guardian see the user as unauthenticated so > filter1 

Try only this request (basic auth)

`1540221802.291          -       172.16.5.15     172.16.5.15     http://prdownloads.sourceforge.net/filezilla/FileZilla_3.2.7.1_win32-setup.exe`

what do you see in squid, nothing ?

**Comment by Schroeffu** [comment]
> what do you see in squid, nothing ?

Yes, that request is not logged in SQUID.
The browser is trying to load the favicon URL, thats the only request which i see in squid/access.log:

tail -f /var/log/e2guardian/access.log /var/log/squid/access.log
```
==> /var/log/e2guardian/access.log <==
1540288276.130          -       172.16.5.15     172.16.5.15     https://www.google.de:443       CONNECT 407     3742    -       -       -       41      -       602     *VERTRAUENSWÜRDIG* Übereinstimmung mit Seite in Ausnahmelistegoogle.de       0       -       InternetAccessLevel0    1
1540288277.240          -       172.16.5.15     172.16.5.15     http://downloads.signotec.com/signoPAD-Tools/signotec_signoPADTools_3.1.9_64Bit.exe     GET     403     0       -       -       -       0       -       900     *VERWEIGERT* Verbotene Datei-Erweiterung: .exe       0       -       InternetAccessLevel0    1

==> /var/log/squid/access.log <==
1540288277.402      0 127.0.0.1 TCP_DENIED/407 4617 GET http://downloads.signotec.com/favicon.ico - HIER_NONE/- text/html

==> /var/log/e2guardian/access.log <==
1540288277.402          -       172.16.5.15     172.16.5.15     http://downloads.signotec.com/favicon.ico       GET     407     4109    -       -       -       0       -       0       *VERTRAUENSWÜRDIG* -    0       -       InternetAccessLevel0 1
```

**Comment by fredbcode** [comment]
Are you using a custom storyboard ?

**Comment by fredbcode** [comment]
You confirm this extract is without NTLM ?

**Comment by Schroeffu** [comment]
> You confirm this extract is without NTLM ?

Yes, I just checked twice. the latest logs above are done with squid basic-auth only, and e2guardian basic-auth module.

> Are you using a custom storyboard ?

No, nothing added there. just the defaults comming with 5.2.2:

- the main config `e2guardian.conf` has inside `preauthstoryboard = '/etc/e2guardian/preauth.story'` 
- and all the files `e2guardianf1/2/3/4.conf` have inside `storyboard = '/etc/e2guardian/examplef1.story'`

in `examplef1.story`everything is commented out, but not the lines:

```
.Include</etc/e2guardian/common.story> 
.Include</etc/e2guardian/site.story>
```

Both files @ pastebin:  
common.story: https://pastebin.com/khT030QV
site.story: https://pastebin.com/z3yGtAVN

the `/etc/e2guardian/preauth.story` has inside:

```
function(pre-authcheck)
if(clientin,bannedclient) return setblock
if(clientin,exceptionclient) return setexception
if(urlin,authexception) setexception
# allow authexceptions to go direct to avoid being blocked by any auth on squid
if(returnset) setgodirect

function(thttps-pre-authcheck)
if(clientin,bannedclient) return setblock
if(clientin,exceptionclient) return setexception
if(urlin,authexception) return setexception


function(icap-pre-authcheck)
```

**Comment by philipianpearce** [comment]
Try changing function(pre-authcheck) in preauth.story to:

function(pre-authcheck)
if(clientin,bannedclient) return setblock
if(clientin,exceptionclient) return setexception
#if(urlin,authexception) setexception
#allow authexceptions to go direct to avoid being blocked by any auth on squid
#if(returnset) setgodirect

If this works it is likely you have these sites in your authexceptionlist(s) so they will be allowed with out checking for user name.

**Comment by Schroeffu** [comment]
Thanks for your reply, i just tried it. 
Unfortunately, it doesn't change the issue. Restarted e2guardian/squid/browsers, made sure BasicAuth is active in Squid+E2Guardian. In Squid, there are multiple BasicAuth Modules, I tried with two of them, it doesn't matter:

- auth_param basic program /usr/bin/ntlm_auth --helper-protocol=squid-2.5-basic
- auth_param basic /usr/lib/squid/basic_ncsa_auth  (...)

My full `preauth.story` :
```
function(pre-authcheck)
if(clientin,bannedclient) return setblock
if(clientin,exceptionclient) return setexception
#if(urlin,authexception) setexception
#allow authexceptions to go direct to avoid being blocked by any auth on squid
#if(returnset) setgodirect

function(thttps-pre-authcheck)
if(clientin,bannedclient) return setblock
if(clientin,exceptionclient) return setexception
if(urlin,authexception) return setexception


function(icap-pre-authcheck)
```

So the http://(.*).exe isn't blocked only when authenticated myself in a 2nd browser-tab with basic auth while opening first HTTPS site (for example github.com). 

EDIT: I just realized a good thing: activated E2Guardian Basic+NTLM Module, and Squid Basic+NTLM Settings, the authentification on HTTP Sites is working after opening any HTTPS Site before. Also without typing the Credentials as BasicAuth, it seems to use the NTLM Auth (and re-use that like basic auth?)

**Comment by Schroeffu** [comment]
Here is a way how to reproduce the issue:

1. Configure Squid with minimal basic auth :

```
auth_param basic program /usr/lib/squid/basic_ncsa_auth /etc/squid/basic_auth_passwords
auth_param basic realm proxy
acl Authenticated_Users proxy_auth REQUIRED
http_access deny !Authenticated_Users all
```

2. Fill the user:password test:test into /etc/squid/basic_auth_passwords: `test:$apr1$9uWCtZb9$VyjtRhdfmjYEiM.qajPJb.` (htpasswd)

3. Make sure basic auth module line in e2guardian is enabled
4. Restart e2guardian, squid, browser
5. open HTTP sites, watch the log: none of them are showing "test" user in the logfile. Browser is not asking for authentication, but on https it does.

For me it looks like it is a general authentication (storyboard, or similar) issue, not only regarding a specific filter. Can reproduce it with Domains listed in `bannedmimetypelist`+`bannedextensionlist` but also `bannedsitelist`

@fredbcode seems really be the same like https://github.com/e2guardian/e2guardian/issues/441 , but there you couln't reproduce it. You may want to try it again?

**Comment by fredbcode** [comment]
You mean

> Browser is not asking for authentication, but on https it does.

Very weird, authentification is not related at all with e2  ! It just watches the request nothing more
This mean that squid doesn't sends the identification request ?  

It works with squid only ? the login pop-up appears or not, the log is right ? If yes do you have a specific acl for e2 IP on squid.conf (127.0.0.1 ?)

**Comment by fredbcode** [comment]
I'm out of office next week, but I will
It should be easy to reproduce

**Comment by PhilippeGCode** [comment]
Hi,

@fredbcode asked me to reproduce your problem

my test : 

I have create a VM with VirtualBox (minimal debian stretch), squid installed with apt-get and e2guardian with the release 5.2.2
The VM is natting with port 3128

e2guardian listen on port 3128 and forward all to 127.0.0.1:3129
squid listen on port 3129

I have follow your config

auth_param basic program /usr/lib/squid/basic_ncsa_auth /etc/squid/basic_auth_passwords
auth_param basic realm proxy
acl Authenticated_Users proxy_auth REQUIRED
http_access deny !Authenticated_Users all

I also created the authentication file with test:test and htpasswd

I update only e2guardian.conf file with the following lines 

filterports = 3128
proxyip = 127.0.0.1
proxyport = 3129

(icap is comment by default so i'm in proxy-chaining mode)


tests made with 
    http://numsys.eu
and
    https://numsys.eu

I have the login pop-up in both case...
(did I forget something about your configuration ?)

**Comment by Schroeffu** [comment]
Until now I am using a bigger Playbook to setup everything (squid, e2guardian, nftables, ldap-utils, ulimit settings, custom CA trust, etc). I will run another test with installing e2guardian/squid manually with minimal settings on a fresh ubuntu 18.04 - similar as you did - for trying to reproduce. 

I need some days, have lot of other things to do beside.

**Comment by Schroeffu** [comment]
@PhilippeGCode thanks again for your time you tried it. Hopefully you still have your VM somewhere :-) Can you try to open a URL with .exe at the end:

- Fresh opened chrome (so no auth popup before)
- Hit this URL directly: http://numsys.eu/blabla.exe

Then you should see the .exe is blocked without asking for authentification before. Open the URL http://numsys.eu - as you did - seems to work properly.

I just created another ubuntu18.04 VM with my playbook and purged e2guardian/squid and reinstalled the default, with the test:test creds above. In my case http://numsys.eu/blabla.exe:

`1541428199.984          -       172.16.5.15     172.16.5.15     http://numsys.eu/blabla.exe     GET     403     0       -       -       -       0       -       900     *DENIED* Banned file extension: .exe    0       -       no_name_group1`

no auth popup :-(

**Comment by fredbcode** [comment]
> |1541428199.984 - 172.16.5.15 172.16.5.15 http://numsys.eu/blabla.exe 
> GET 403 0 - - - 0 - 900 *DENIED* Banned file extension: .exe 0 - 
> no_name_group1|
>
> no auth popup :-(
>
>

It look like group1 rules are applied before authentification

**Comment by fredbcode** [comment]
@Schroeffu

Can you try with the original .story files ? I saw in your comment 
before some changes

https://github.com/e2guardian/e2guardian/tree/v5.2/configs

Philippe, tu peux tester ?

**Comment by fredbcode** [comment]
@philipianpearce @PhilippeGCode there is something wrong here, yes @Schroeffu is right if something is denied for group1 the user is blocked before his identification

**Comment by philipianpearce** [comment]
I think this is a long-standing issue, not specific to v5. I.e This was the way it worked in previous releases? 

I have an idea on how it may be fixed, but this requires some changes to the overall logic and an extra call to squid. When I get a chance I will try and code it in a temporary branch. 

Philip 

----- Original Message -----

**Comment by fredbcode** [comment]
Hi,

We will try to reproduce from a previous version, but I'm pretty sure 
that it works.

We had 6 groups with the first blocked for all requests, and no 
complaint from users about identification

**Comment by PhilippeGCode** [comment]
Hi,

I have test with v4

The website "http://no-ip.org" is in e2guardian's black-list

The test:

(Request)
wget http://no-ip.org
--2018-11-07 16:31:22--  http://no-ip.org/
Connexion à 192.168.0.1:3128… connecté.
requête Proxy transmise, en attente de la réponse… 407 Proxy Authentication Required

(It's in french but just understand the auth popup will spawn)

(Log)
cat /var/log/squid/access.log | grep "no-ip.org"
192.168.0.1 - - [07/Nov/2018:16:31:22 +0100] "GET http://no-ip.org/ HTTP/1.0" 407 TCP_DENIED:HIER_NONE "Wget/1.16 (linux-gnu)"

So it's v5 problem

**Comment by philipianpearce** [comment]
@Schroeffu   Would you test with temporary pxauth branch?    If you are happy with it I will then roll into v5.2.

**Comment by PhilippeGCode** [comment]
[debug_pxauth.txt](https://github.com/e2guardian/e2guardian/files/2599210/debug_pxauth.txt)
Hi,

I have tested and It doesn't work 

my test :
(I reused my VM so I will just describe the e2guardian part)

apt-get update && apt-get upgrade -y && apt-get install git -y && apt-get build-dep e2guardian
git clone https://github.com/e2guardian/e2guardian.git
cd e2guardian && git checkout pxauth

./autogen.sh && ./configure '--prefix=/usr' '--enable-clamd=yes' '--with-proxyuser=e2guardian' '--with-proxygroup=e2guardian' '--sysconfdir=/etc' '--localstatedir=/var' '--enable-icap=yes' '--enable-commandline=yes' '--enable-email=yes' '--enable-ntlm=yes' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--enable-pcre=yes' '--enable-sslmitm=yes' 'CPPFLAGS=-mno-sse2 -g -O2' '--with-dgdebug=on' && make

killall e2guardian
src/e2guardian -N > debug_pxauth.txt 2>&1

In attach file the debug_pxauth.txt

**Comment by philipianpearce** [comment]
@PhilippeGCode   
Would you check that you have the proxy authplugin enabled in e2guardian.conf?     Dump indicates that they are not - hw5:  -No auth plugins loaded; using defaults & feigning persistency
Thanks
Philip

**Comment by philipianpearce** [comment]
I have now tested pxauth with basic-auth and seems to working correctly.  I would still like OK tests from others before merging this branch.

**Comment by PhilippeGCode** [comment]
I have tested and it works :)

The account of user appears too in the log file

**Comment by philipianpearce** [comment]
Thanks for testing.   Now merged into v5.2 and v5develop.
Closing issue.
